pdal/private/Raster.hpp

```diff
@@ -0,0 +1,172 @@
+         m_widthArg = &args.add("width", "Number of cells in the X direction", width);
+         m_heightArg = &args.add("height", "Number of cells in the Y direction", height);
+    }
+
+    int checkArgs()
+    {
+        int args = 0;
+        if (m_xOriginArg->set())
+            args++;
+        if (m_yOriginArg->set())
+            args++;
+        if (m_widthArg->set())
+            args++;
+        if (m_heightArg->set())
+            args++;
+        return args;
+    }
+
+public:
+    double xOrigin;
+    double yOrigin;
+    size_t width;
+    size_t height;
+    double edgeLength;
+
+private:
+    Arg *m_xOriginArg;
+    Arg *m_yOriginArg;
+    Arg *m_widthArg;
+    Arg *m_heightArg;
+};
+
+// This is a raw raster suitable for sticking into GDAL. X goes right to left, Y goes from
+// top to bottom.
+template <typename T>
+class Raster
+{
+public:
+    using iterator = typename std::vector<T>::iterator;
+    using const_iterator = typename std::vector<T>::const_iterator;
+
+    Raster(const T& initializer = T()) : m_initializer(initializer)
+    {}
+
+    Raster(const RasterLimits& limits, const T& initializer = T()) :
+        m_limits(limits), m_data(width() * height(), initializer), m_initializer(initializer)
+    {}
+
+    size_t width() const
+        { return m_limits.width; }
+
+    size_t height() const
+        { return m_limits.height; }
+
+    double edgeLength() const
+        { return m_limits.edgeLength; }
+
+    double xOrigin() const
+        { return m_limits.xOrigin; }
+
+    double yOrigin() const
+        { return m_limits.yOrigin; }
+
+    size_t index(size_t i, size_t j) const
+        { return (j * width()) + i; }
+
+    int horizontalIndex(double x) const
+        { return (int)(x / edgeLength()); }
+
+    int verticalIndex(double y) const
+        { return height() - (int)(y / edgeLength()) - 1; }
+
+    double horizontalPos(size_t i) const
+        { return (i + .5) * edgeLength(); }
+
+    double verticalPos(size_t j) const
+        { return (height() - (j + .5)) * edgeLength(); }
+
+    const T *data() const
+        { return m_data.data(); }
+
+    T *data()
+        { return m_data.data(); }
+
+    const T& at(size_t x, size_t y) const
+        { return m_data[index(x, y)]; }
+
+    T& at(size_t x, size_t y)
+        { return m_data[index(x, y)]; }
+
+    iterator begin()
+        { return m_data.begin(); }
+
+    const_iterator begin() const
+        { return m_data.begin(); }
+
+    iterator end()
+        { return m_data.end(); }
+
+    const_iterator end() const
+        { return m_data.end(); }
+
+    void setLimits(const RasterLimits& limits)
+    {
+        m_limits = limits;
+        m_data.swap(std::vector<T>(width() * height(), m_initializer));
+    }
+
+    const RasterLimits& limits() const
+        { return m_limits; }
+
+private:
+    RasterLimits m_limits;
+    std::vector<T> m_data;
+    T m_initializer;
+};
+
+using Rasterd = Raster<double>;
```

Comment:
Totally bikeshedding, but I don't like `Rasterd`. Maybe `RasterD` is better? Doesn't really matter in the end.

Why `using` and not a `typedef`?

Reply:
I was following the Eigen way on the name. I don't care.

I like `using` better. It works in template cases where `typedef` doesn't. And if you declare function pointers as types, the `typedef` way is yuck. That's all I got.

Reply:
I guess you are right, it's also lower case in Eigen, just separated by the dimensionality. It just looked at first like a typo when I first encountered it in the FaceRasterFilter code.

:+1: to `using`. It just caught me off guard and we use plenty of `typedefs` elsewhere hence the curiosity. No need to change.